e2e/aws/rds_test.go

```diff
@@ -100,186 +113,258 @@ func awsDBDiscoveryUnmatched(t *testing.T) {
 	}
-	t.Run("connection", func(t *testing.T) {
-		tests := []struct {
-			name             string
-			route            tlsca.RouteToDatabase
-			testDBConnection dbConnectionTestFunc
-		}{
-			{
-				name:             "RDS postgres instance",
-				route:            rdsPostgresInstance,
-				testDBConnection: postgresConnTestFn(cluster),
-			},
-			{
-				name:             "RDS postgres instance via local proxy",
-				route:            rdsPostgresInstance,
-				testDBConnection: postgresLocalProxyConnTestFn(cluster),
-			},
-			{
-				name:             "RDS MySQL instance via local proxy",
-				route:            rdsMySQLInstance,
-				testDBConnection: mySQLLocalProxyConnTestFn(cluster),
-			},
-		}
-		for _, test := range tests {
-			test := test
-			t.Run(test.name, func(t *testing.T) {
-				t.Parallel()
-				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
-				defer cancel()
-				test.testDBConnection(t, ctx, test.route)
-			})
-		}
+	cluster := makeDBTestCluster(t, accessRole, discoveryRole, types.AWSMatcherRDS, opts...)
+
+	t.Run("postgres", func(t *testing.T) {
+		t.Parallel()
+
+		// wait for the database to be discovered
+		pgDBName := mustGetEnv(t, rdsPostgresInstanceNameEnv)
+		waitForDatabases(t, cluster.Process, pgDBName)
+		db, err := cluster.Process.GetAuthServer().GetDatabase(ctx, pgDBName)
+		require.NoError(t, err)
+		adminUser := mustGetDBAdmin(t, db)
+
+		conn := connectAsRDSPostgresAdmin(t, ctx, db.GetAWS().RDS.InstanceID)
+		provisionRDSPostgresAutoUsersAdmin(t, ctx, conn, adminUser.Name)
+		t.Cleanup(func() {
+			// best effort cleanup all the users created for the tests,
+			// including the auto drop user in case Teleport fails to do so.
+			_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserKeep))
+			_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserDrop))
+		})
+
+		t.Run("connect", func(t *testing.T) {
+			for name, test := range map[string]struct {
+				user   string
+				dbUser string
+			}{
+				"existing user":  {user: hostUser, dbUser: adminUser.Name},
+				"auto user keep": {user: autoUserKeep, dbUser: autoUserKeep},
+				"auto user drop": {user: autoUserDrop, dbUser: autoUserDrop},
+			} {
+				test := test
+				route := tlsca.RouteToDatabase{
+					ServiceName: db.GetName(),
+					Protocol:    defaults.ProtocolPostgres,
+					Username:    test.dbUser,
+					Database:    "postgres",
+				}
+				t.Run(name+"/via proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresConnTest(t, cluster, test.user, route)
+				})
+				t.Run(name+"/via local proxy", func(t *testing.T) {
+					t.Parallel()
+					postgresLocalProxyConnTest(t, cluster, test.user, route)
+				})
+			}
+		})
+
+		t.Run("connect/auto user keep", func(t *testing.T) {
+			waitForPostgresAutoUserDeactivate(t, ctx, conn, autoUserKeep)
+		})
+		t.Run("connect/auto user drop", func(t *testing.T) {
+			waitForPostgresAutoUserDrop(t, ctx, conn, autoUserDrop)
+		})
```

Comment:
The cleanup tests are done after all the connections have finished. I did all connection tests in parallel not only for test speed but also to put some parallel connections under test.

I find it useful to name these cleanup tests like I have because I can run `go test -run "connect/auto_user_drop"` and it runs all the auto user drop connection tests + cleanup test.
If I named it like `"auto user drop cleanup"` then `go test -run "auto_user_drop_cleanup"` would fail because it doesn't run the connection tests.

Reply:
i rearranged the subtests so i don't have to do this naming hack